File: cpp/src/registration.cpp
#include "acpd/analytic.hpp"
#include "acpd/rigid.hpp"
#include <algorithm>
#include <cmath>
#include <memory>
namespace acpd {
    namespace {
        void world_units(Stage& s,double scale) {
            s.initial_sigma2*=scale*scale;
            s.final_sigma2*=scale*scale;
            for(auto& h:s.history) {
                h.sigma2*=scale*scale;
                h.step_rms*=scale;
                h.fit_rms*=scale;
            }
        }
        double motion_rms(const Matrix& a,const Matrix& b) {
            return (a-b).norm()/std::sqrt(double(a.rows()));
        }
    }
    Result registration(const Matrix& fixed,const Matrix& moving,const Options& o,const Matrix& r0,const Vector& t0,const Matrix& normals) {
        validate_pair(fixed,moving,true);
        o.validate();
        const int d=static_cast<int>(fixed.cols());
        validate_pose(r0,t0,d);
        if(normals.size()) validate_normals(normals,fixed);
        if(o.method!=Method::Analytic && o.rigid.objective=="point_to_plane" && !normals.size())
        throw std::invalid_argument("point-to-plane requires target_normals");
        const Vector center=fixed.colwise().mean();
        Matrix x=fixed.rowwise()-center.transpose(),y0=moving.rowwise()-center.transpose();
        const double scale=x.norm()/std::sqrt(double(x.rows()));
        if(!std::isfinite(scale)||scale<=1e-150) throw NumericalError("fixed cloud has zero or unrepresentable extent");
        x/=scale;
        y0/=scale;
        require_finite(x,"normalization");
        require_finite(y0,"normalization");
        Matrix rotation=r0;
        Vector translation=(r0*center+t0-center)/scale;
        Matrix y=apply_pose(y0,rotation,translation);
        Result out;
        out.center=center;
        out.normalization_scale=scale;
        out.method=o.method;
        out.backend=o.backend;
        double sigma2=0;
        if(o.method!=Method::Analytic) {
            Stage& stage=out.rigid_stage;
            sigma2=o.rigid.sigma2>0?o.rigid.sigma2/scale/scale:initial_variance(x,y);
            sigma2=std::max(sigma2,o.rigid.min_sigma2);
            stage.initial_sigma2=sigma2;
            stage.stop_reason="iteration_limit";
            std::unique_ptr<FixedNoBlurLattice> cache;
            double indexed_sigma=-1;
            const Matrix nv=o.rigid.objective=="point_to_plane"?normals:Matrix();
            for(int it=0;it<o.rigid.max_iterations;++it) {
                if(o.backend==Backend::PermutohedralNoBlur && (!cache||sigma2!=indexed_sigma)) {
                    cache=std::make_unique<FixedNoBlurLattice>(x/std::sqrt(sigma2),moment_values(x,nv));
                    indexed_sigma=sigma2;
                    ++stage.index_builds;
                }
                const Statistics stats=posterior_statistics(x,y,sigma2,o.rigid.w,true,o.backend,nv,cache.get(),o.fgt);
                if(o.backend==Backend::Permutohedral||o.backend==Backend::Fgt) ++stage.index_builds;
                if(o.backend==Backend::Probreg) stage.index_builds+=(stats.lattice_mode=="probreg_noblur"?2:1);
                const RigidFit fit=fit_rigid(y,stats,o.rigid);
                // Use UPDATED coordinates and ambient dimension d. The attached
                // probreg.py uses old coordinates and literal 3; see SOURCE_DIFFERENCES.md.
                const double next_sigma=o.rigid.update_sigma2?variance_from_statistics(fit.next,stats,o.rigid.min_sigma2):sigma2;
                const double motion=motion_rms(y,fit.next),change=std::abs(next_sigma-sigma2)/(sigma2+1e-12);
                rotation=fit.rotation*rotation;
                translation=fit.rotation*translation+fit.translation;
                y=fit.next;
                stage.history.push_back({
                    it+1,0,0,fit.active,fit.rank,next_sigma,stats.nll,motion,fit.fit_rms,stats.vertices,stats.lattice_mode
                });
                sigma2=next_sigma;
                stage.best_iteration=it+1;
                if(motion<=o.rigid.tolerance&&change<=o.rigid.tolerance) {
                    stage.converged=true;
                    stage.stop_reason="tolerance";
                    break;
                }
            }
            stage.final_sigma2=sigma2;
        }
        // The rigid pose is immutable below this line; residual maps are composed
        // AFTER it, not absorbed back into rotation/translation.
        out.rotation=rotation;
        out.translation=scale*translation+center-rotation*center;
        out.rigid_transformed=apply_pose(moving,out.rotation,out.translation);
        if(o.method!=Method::Rigid) {
            const auto& opt=o.analytic;
            Stage& stage=out.analytic_stage;
            // "auto" is resolved from the declared method, never from a runtime quality
            // test. Nonrigid inherits the variance the rigid stage annealed to; standalone
            // analytic uses the ACPD/CPD all-pair initialization of the paper.
            const std::string initialization=opt.initialization=="auto"
            ?(o.method==Method::Nonrigid?std::string("filterreg"):std::string("cpd")):opt.initialization;
            if(opt.sigma2>0) sigma2=opt.sigma2/scale/scale;
            else if(initialization=="cpd") sigma2=initial_variance(x,y);
            sigma2=std::max(sigma2,opt.min_sigma2);
            stage.initial_sigma2=sigma2;
            stage.stop_reason="iteration_limit";
            Matrix best_y=y;
            double best_sigma=sigma2,best_score=std::sqrt(d*sigma2),previous_score=best_score;
            std::size_t best_steps=0;
            int stable=0,no_improve=0,previous_degree=-1;
            const auto schedule=degree_schedule(opt.max_iterations,opt.min_degree,opt.max_degree);
            // Degree continuation: the schedule bounds how long each degree MAY run,
            // it does not require running that long. Convergence at a degree below
            // max_degree advances the continuation instead of ending the whole stage;
            // only the highest scheduled degree may terminate it. See
            // docs/SOURCE_DIFFERENCES.md, Analytic-CPD 10.
            // Dimensionless: the fixed cloud is normalized to unit RMS radius, and the
            // limit tracks its own largest radius so an outlier-heavy cloud is not
            // penalised. See the divergence_radius note in types.hpp.
            const double divergence_limit=opt.divergence_radius*std::max(1.0,x.rowwise().norm().maxCoeff());
            std::size_t cursor=0;
            const auto advance_degree=[&schedule](std::size_t at)->std::size_t {
                const int degree=schedule[at];
                while(at<schedule.size()&&schedule[at]<=degree) ++at;
                return at;
            };
            for(int it=0;it<opt.max_iterations&&cursor<schedule.size();++it,++cursor) {
                try {
                    const Statistics stats=posterior_statistics(x,y,sigma2,opt.w,false,opt.backend,Matrix(),nullptr,o.fgt);
                    int active=0;
                    for(int i=0;i<stats.rho.size();++i) if(stats.rho[i]>opt.min_mass) ++active;
                    if(stats.mass<=opt.min_mass||active<static_cast<int>(exponents(d,opt.min_degree).size())) {
                        stage.stop_reason="insufficient_posterior_mass";
                        break;
                    }
                    const Fit fit=fit_analytic(y,stats,schedule[cursor],opt);
                    if(fit.next.rowwise().norm().maxCoeff()>divergence_limit) {
                        // The analytic map is a global polynomial constrained only where the
                        // posterior supports it. Once a point loses support it can be
                        // extrapolated arbitrarily far, and the rho-weighted variance cannot
                        // see it, so a diverging state could otherwise be recorded as best.
                        // Refuse the iterate; the fit itself is left exactly as computed.
                        stage.stop_reason="numerical_divergence";
                        break;
                    }
                    if(previous_degree>=0&&previous_degree!=fit.step.degree) {
                        stable=0;
                        no_improve=0;
                    }
                    previous_degree=fit.step.degree;
                    const double next_sigma=variance_from_statistics(fit.next,stats,opt.min_sigma2),score=std::sqrt(d*next_sigma);
                    const double delta_y=(fit.next-y).norm()/(y.norm()+1e-12);
                    const double delta_sigma=std::abs(next_sigma-sigma2)/(std::abs(sigma2)+1e-12);
                    const double delta_score=std::abs(score-previous_score)/(std::abs(previous_score)+1e-12);
                    const double motion=motion_rms(y,fit.next);
                    const bool significant=best_score-score>std::max(1e-12,opt.improvement_relative*std::abs(best_score));
                    no_improve=significant?0:no_improve+1;
                    out.steps.push_back(fit.step);
                    y=fit.next;
                    sigma2=next_sigma;
                    stage.history.push_back({
                        it+1,schedule[cursor],fit.step.degree,fit.active,fit.rank,sigma2,stats.nll,motion,fit.fit_rms,
                            stats.vertices,stats.lattice_mode
                    });
                    // Fig.1: save the actual best state. Patience uses the source's
                    // significant-improvement threshold, but never loses a lower score.
                    if(score<best_score) {
                        best_score=score;
                        best_y=y;
                        best_sigma=sigma2;
                        best_steps=out.steps.size();
                        stage.best_iteration=it+1;
                    }
                    stable=(delta_y<opt.tolerance&&delta_sigma<opt.tolerance&&delta_score<opt.tolerance)?stable+1:0;
                    previous_score=score;
                    if(score<opt.tolerance) {
                        stage.converged=true;
                        stage.stop_reason="residual_tolerance";
                        break;
                    }
                    if(it+1>=opt.min_iterations) {
                        const bool is_stable=stable>=opt.stable_patience;
                        const bool is_rebound=no_improve>=opt.stable_patience&&score>best_score*(1+opt.rebound_relative)+1e-12;
                        const bool is_stalled=no_improve>=opt.no_improve_patience;
                        if(is_stable||is_rebound||is_stalled) {
                            const std::size_t next=advance_degree(cursor);
                            if(next<schedule.size()) {
                                // Converged at this degree with higher degrees still scheduled:
                                // continue the sequence rather than reporting the whole stage done.
                                cursor=next-1;
                                stable=0;
                                no_improve=0;
                                previous_degree=-1;
                                continue;
                            }
                            if(is_stable) {
                                stage.converged=true;
                                stage.stop_reason="stable_tolerance";
                            }
                            else stage.stop_reason=is_rebound?"internal_rebound":"no_improvement";
                            break;
                        }
                    }
                }
                catch(const NumericalError&) {
                    // The analytic map is unregularized by design, so an ill-posed pair can
                    // drive the Taylor basis or the second-moment residual out of range. Report
                    // it as a stop reason and return the best state that was actually
                    // evaluated; never a silently substituted or re-tuned computation. A
                    // failure before any iteration succeeded is a genuine input error and
                    // still propagates.
                    if(stage.history.empty()) throw;
                    stage.stop_reason="numerical_divergence";
                    break;
                }
            }
            y=best_y;
            sigma2=best_sigma;
            out.steps.resize(best_steps);
            stage.final_sigma2=sigma2;
        }
        out.sigma2=sigma2*scale*scale;
        out.transformed=y*scale;
        out.transformed.rowwise()+=center.transpose();
        require_finite(out.transformed,"result");
        if(!std::isfinite(out.sigma2)) throw NumericalError("world-coordinate variance overflow");
        world_units(out.rigid_stage,scale);
        world_units(out.analytic_stage,scale);
        return out;
    }
}
// namespace acpd

File: rust/crates/acpd-core/src/registration.rs
//! Two-stage registration. The rigid pose is frozen before the analytic stage.
use crate::analytic::{
    degree_schedule, exponents, fit_analytic
};
use crate::gaussian::{
    initial_variance, moment_values, posterior_statistics, variance_from_statistics
};
use crate::lattice::FixedNoBlurLattice;
use crate::rigid::fit_rigid;
use crate::types::*;
fn world_units(stage: &mut Stage, scale: f64) {
    stage.initial_sigma2 *= scale*scale;
    stage.final_sigma2 *= scale*scale;
    for h in &mut stage.history {
        h.sigma2 *= scale*scale;
        h.step_rms *= scale;
        h.fit_rms *= scale;
    }
}
fn motion_rms(a: &Matrix,b: &Matrix) -> f64 {
    (a-b).norm()/(a.nrows() as f64).sqrt()
}
pub fn registration(
fixed: &Matrix, moving: &Matrix, options: &Options,
initial_rotation: &Matrix, initial_translation: &Vector,
target_normals: Option<&Matrix>,
) -> RegResult<RegistrationResult> {
    validate_pair(fixed,moving,true)?;
    options.validate()?;
    let d = fixed.ncols();
    validate_pose(initial_rotation,initial_translation,d)?;
    if let Some(normals) = target_normals {
        validate_normals(normals,fixed)?;
    }
    if options.method != Method::Analytic && options.rigid.objective == "point_to_plane" && target_normals.is_none() {
        return Err(invalid("point-to-plane requires target_normals"));
    }
    let center = mean(fixed);
    let mut x = Matrix::from_fn(fixed.nrows(),d,|i,a| fixed[(i,a)]-center[a]);
    let mut y0 = Matrix::from_fn(moving.nrows(),d,|i,a| moving[(i,a)]-center[a]);
    let scale = x.norm()/(x.nrows() as f64).sqrt();
    if !scale.is_finite() || scale <= 1e-150 {
        return Err(numerical("fixed cloud has zero or unrepresentable extent"));
    }
    x /= scale;
    y0 /= scale;
    finite(&x,"normalization")?;
    finite(&y0,"normalization")?;
    let mut rotation = initial_rotation.clone();
    let mut translation = (initial_rotation*&center+initial_translation-&center)/scale;
    let mut y = apply_pose(&y0,&rotation,&translation)?;
    let mut sigma2 = 0.0;
    let mut rigid_stage = Stage::default();
    let mut analytic_stage = Stage::default();
    let mut steps = Vec::new();
    if options.method != Method::Analytic {
        let opt = &options.rigid;
        sigma2 = if opt.sigma2 > 0.0 {
            opt.sigma2/scale/scale
        }else {
            initial_variance(&x,&y)?
        };
        sigma2 = sigma2.max(opt.min_sigma2);
        rigid_stage.initial_sigma2 = sigma2;
        rigid_stage.stop_reason = "iteration_limit".into();
        let mut cache: Option<FixedNoBlurLattice> = None;
        let mut indexed_sigma = -1.0;
        let normals = if opt.objective == "point_to_plane" {
            target_normals
        }else {
            None
        };
        for it in 0..opt.max_iterations {
            if options.backend == Backend::PermutohedralNoBlur && (cache.is_none() || sigma2 != indexed_sigma) {
                cache = Some(FixedNoBlurLattice::new(&(&x/sigma2.sqrt()),&moment_values(&x,normals)?)?);
                indexed_sigma = sigma2;
                rigid_stage.index_builds += 1;
            }
            let stats = posterior_statistics(&x,&y,sigma2,opt.w,true,options.backend,normals,cache.as_ref())?;
            if options.backend == Backend::Permutohedral {
                rigid_stage.index_builds += 1;
            }
            if options.backend == Backend::Probreg {
                rigid_stage.index_builds += if stats.lattice_mode == "probreg_noblur" {
                    2
                }else {
                    1
                };
            }
            let fit = fit_rigid(&y,&stats,opt)?;
            // Updated coordinates and ambient dimension, not the old-coordinate
            // / literal-three update found in the attached probreg Python code.
            let next_sigma = if opt.update_sigma2 {
                variance_from_statistics(&fit.next,&stats,opt.min_sigma2)?
            }else {
                sigma2
            };
            let motion = motion_rms(&y,&fit.next);
            let change = (next_sigma-sigma2).abs()/(sigma2+1e-12);
            rotation = &fit.rotation*rotation;
            translation = &fit.rotation*translation+&fit.translation;
            y = fit.next;
            rigid_stage.history.push(Iteration {
                iteration:it+1,raw_degree:0,degree:0,active:fit.active,rank:fit.rank,
                sigma2:next_sigma,nll_before:stats.nll,step_rms:motion,fit_rms:fit.fit_rms,
                lattice_vertices:stats.vertices,lattice_mode:stats.lattice_mode,
            });
            sigma2 = next_sigma;
            rigid_stage.best_iteration = it+1;
            if motion <= opt.tolerance && change <= opt.tolerance {
                rigid_stage.converged = true;
                rigid_stage.stop_reason = "tolerance".into();
                break;
            }
        }
        rigid_stage.final_sigma2 = sigma2;
    }
    // No assignment to rotation or world_translation after this point.
    let world_translation = scale*&translation+&center-&rotation*&center;
    let rigid_transformed = apply_pose(moving,&rotation,&world_translation)?;
    if options.method != Method::Rigid {
        let opt = &options.analytic;
        if opt.sigma2 > 0.0 {
            sigma2 = opt.sigma2/scale/scale;
        }
        else if opt.initialization == "cpd" {
            sigma2 = initial_variance(&x,&y)?;
        }
        sigma2 = sigma2.max(opt.min_sigma2);
        analytic_stage.initial_sigma2 = sigma2;
        analytic_stage.stop_reason = "iteration_limit".into();
        let mut best_y = y.clone();
        let mut best_sigma = sigma2;
        let mut best_score = (d as f64*sigma2).sqrt();
        let mut previous_score = best_score;
        let mut best_steps = 0;
        let mut stable = 0;
        let mut no_improve = 0;
        let mut previous_degree: Option<usize> = None;
        let schedule = degree_schedule(opt.max_iterations,opt.min_degree,opt.max_degree)?;
        for (it,&raw_degree) in schedule.iter().enumerate() {
            let stats = posterior_statistics(&x,&y,sigma2,opt.w,false,Backend::Direct,None,None)?;
            let active = stats.rho.iter().filter(|&&v| v>opt.min_mass).count();
            if stats.mass <= opt.min_mass || active < exponents(d,opt.min_degree)?.len() {
                analytic_stage.stop_reason = "insufficient_posterior_mass".into();
                break;
            }
            let fit = fit_analytic(&y,&stats,raw_degree,opt)?;
            if previous_degree.is_some_and(|degree| degree != fit.step.degree) {
                stable = 0;
                no_improve = 0;
            }
            previous_degree = Some(fit.step.degree);
            let next_sigma = variance_from_statistics(&fit.next,&stats,opt.min_sigma2)?;
            let score = (d as f64*next_sigma).sqrt();
            let delta_y = (&fit.next-&y).norm()/(y.norm()+1e-12);
            let delta_sigma = (next_sigma-sigma2).abs()/(sigma2.abs()+1e-12);
            let delta_score = (score-previous_score).abs()/(previous_score.abs()+1e-12);
            let motion = motion_rms(&y,&fit.next);
            let significant = best_score-score > 1e-12_f64.max(opt.improvement_relative*best_score.abs());
            no_improve = if significant {
                0
            }else {
                no_improve+1
            };
            let degree = fit.step.degree;
            steps.push(fit.step);
            y = fit.next;
            sigma2 = next_sigma;
            analytic_stage.history.push(Iteration {
                iteration:it+1,raw_degree,degree,active:fit.active,rank:fit.rank,
                sigma2,nll_before:stats.nll,step_rms:motion,fit_rms:fit.fit_rms,
                lattice_vertices:0,lattice_mode:"direct".into(),
            });
            // Actual minimum is retained; significant improvement controls only patience.
            if score < best_score {
                best_score = score;
                best_y = y.clone();
                best_sigma = sigma2;
                best_steps = steps.len();
                analytic_stage.best_iteration = it+1;
            }
            stable = if delta_y<opt.tolerance && delta_sigma<opt.tolerance && delta_score<opt.tolerance {
                stable+1
            }else {
                0
            };
            previous_score = score;
            if score < opt.tolerance {
                analytic_stage.converged = true;
                analytic_stage.stop_reason = "residual_tolerance".into();
                break;
            }
            if it+1 >= opt.min_iterations {
                if stable >= opt.stable_patience {
                    analytic_stage.converged = true;
                    analytic_stage.stop_reason = "stable_tolerance".into();
                    break;
                }
                if no_improve >= opt.stable_patience && score > best_score*(1.0+opt.rebound_relative)+1e-12 {
                    analytic_stage.stop_reason = "internal_rebound".into();
                    break;
                }
                if no_improve >= opt.no_improve_patience {
                    analytic_stage.stop_reason = "no_improvement".into();
                    break;
                }
            }
        }
        y = best_y;
        sigma2 = best_sigma;
        steps.truncate(best_steps);
        analytic_stage.final_sigma2 = sigma2;
    }
    let transformed = Matrix::from_fn(y.nrows(),d,|i,a| y[(i,a)]*scale+center[a]);
    finite(&transformed,"result")?;
    if !(sigma2*scale*scale).is_finite() {
        return Err(numerical("world-coordinate variance overflow"));
    }
    world_units(&mut rigid_stage,scale);
    world_units(&mut analytic_stage,scale);
    Ok(RegistrationResult {
        rotation,translation:world_translation,center,normalization_scale:scale,
        transformed,rigid_transformed,steps,rigid_stage,analytic_stage,
        sigma2:sigma2*scale*scale,method:options.method,backend:options.backend,
    })
}
